include/morphio/collection.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include <morphio/morphology.h>
+#include <morphio/mut/morphology.h>
+
+namespace morphio {
+
+class CollectionImpl
+{
+  public:
+    virtual ~CollectionImpl() = default;
+
+    virtual Morphology load(const std::string& morph_name) const = 0;
+    virtual mut::Morphology load_mut(const std::string& morph_name) const = 0;
+};
+
+class Collection
+{
+  public:
+    Collection(std::shared_ptr<CollectionImpl> collection);
+
+    /**
+     * Create a collection from the given path.
+     *
+     * If `collection_path` points to an HDF5 file, then that file
+     * must be a container. Otherwise the `collection_path` should
+     * point to the directory containing the morphology files.
+     *
+     * If the collection path is a directory, the extension of the morphology
+     * file must be guessed. The optional argument `extensions` specifies which
+     * and in which order the morphologies are searched.
+     */
+    Collection(std::string collection_path,
+               std::vector<std::string> extensions = std::vector<std::string>{"h5", "swc", "asc"});
+
+    /**
+     * Load the morphology as an immutable morphology.
+     */
+    template <class M>
+    typename std::enable_if<std::is_same<M, Morphology>::value, M>::type load(
+        const std::string& morph_name) const {
+        if (_collection != nullptr) {
+            return _collection->load(morph_name);
+        }
+
+        throw std::runtime_error("The collection has been closed.");
+    }
+
+    /**
+     * Load the morphology as a mutable morphology.
+     */
+    template <class M>
+    typename std::enable_if<std::is_same<M, mut::Morphology>::value, M>::type load(
+        const std::string& morph_name) const {
+        if (_collection != nullptr) {
+            return _collection->load_mut(morph_name);
+        }
+
+        throw std::runtime_error("The collection has been closed.");
+    }
+
+    /**
+     * Close the collection.
+     *
+     * Note that `morphio::Collection` uses RAII. Therefore, the usual
+     * scoping rules should suffice. However, there are cases where one
+     * wants to close the collection explicitly, causing all resources held
+     * by the collection to be release. In the case of containers it ensures
+     * that the file is closed.
+     *
+     * The object should not be used after calling `close`.
+     */
+    void close();
+
```

Comment:
I wonder if a useful addition would be to have `list` or `contents` to get the available morphologies within the container?

Reply:
Good idea! I was thinking we'd get the morphology names from libsonata via a nodesfile; but that's not always the case. For directories it might be difficult to distinguish between files that contain morphologies and files that happen to have an extension that's used for morphologies. This gets slightly murkier since we often(?) nest morphologies in subdirectories of the morphology directory. The subdirectory then becomes part of the morphology name.

Reply:
@mgeplf Can this please be a separate task? I'd like to prepare the migration for two of our internal applications; but their CI will only run after this has been merged. Therefore, it would be helpful to limit the functionality of this PR.

So far I've not needed know what's inside a collection, I've always had a nodesfile tell me what I need to load.

Reply:
yeah, that's no problem

We'll need a way to list what's there, at least of h5 files, so that we can quickly reconcile that with nodes files, if things aren't working - but we can do that in the next step.